contracts/Removal.sol

```diff
@@ -173,126 +234,144 @@ contract Removal is
-    // todo do we add any validation to enforce that all removals in batch belong to the same project id?
-    BatchMintRemovalsData memory decodedData = abi.decode(
-      data,
-      (BatchMintRemovalsData)
-    );
-    uint256 projectId = decodedData.projectId;
-    uint256 scheduleStartTime = decodedData.scheduleStartTime;
-    uint256 holdbackPercentage = decodedData.holdbackPercentage;
-    bool list = decodedData.list;
-    address marketAddress = decodedData.marketAddress;
-    if (projectId == 0) {
-      revert InvalidProjectId({projectId: projectId});
+    uint256[] memory totals = balanceOfBatch(owners, ids);
+    uint256 total = 0;
+    for (uint256 i = 0; i < totals.length; ++i) {
+      total += totals[i];
     }
-    uint256 newTokenIdCounter = tokenIdCounter;
-    for (uint256 i = 0; i < idsLength; i++) {
-      uint256 id = ids[i];
-      if (_tokenIdExists[id]) {
-        revert TokenIdExists({tokenId: id});
-      }
-      _removalIdToRemovalData[id].projectId = projectId;
-      _removalIdToRemovalData[id].holdbackPercentage = holdbackPercentage;
+    return total;
+  }
+
+  // todo this function will not scale well- consider dropping it somehow
+  function tokensOfOwner(
+    address owner // todo global rename (tokens -> removals?)
+  ) external view returns (uint256[] memory) {
+    return _addressToOwnedTokenIds[owner].values();
+  }
 
-      _tokenIdExists[id] = true;
-      indexToTokenId[newTokenIdCounter] = id;
-      newTokenIdCounter += 1;
+  // todo rename cumulativeBalanceOfOwner
+  // todo this function will not scale well- consider dropping it somehow
+  function cumulativeBalanceOf(address owner) external view returns (uint256) {
+    // todo are the bodies of these functions re-usable across the contract? Seems like an abstraction might be possible
+    EnumerableSetUpgradeable.UintSet storage removals = _addressToOwnedTokenIds[
+      owner
+    ];
+    uint256 numberOfTokensOwned = removals.length();
+    address[] memory owners = new address[](numberOfTokensOwned);
+    for (uint256 i = 0; i < numberOfTokensOwned; ++i) {
+      owners[i] = owner;
     }
-    uint256 firstId = ids[0];
-    tokenIdCounter = newTokenIdCounter;
-    _projectIdToScheduleData[projectId] = ScheduleData(
-      scheduleStartTime,
-      firstId.supplierAddress(), // todo check gas costs of extracting values per line vs once using unpackRemovalIdV0
-      firstId.methodology(), // todo check gas costs of extracting values per line vs once using unpackRemovalIdV0
-      firstId.methodologyVersion() // todo check gas costs of extracting values per line vs once using unpackRemovalIdV0
-    );
-    super._mintBatch(to, ids, amounts, data);
-    setApprovalForAll(to, _msgSender(), true); // todo look at vesting contract for potentially better approach
-    if (list) {
-      safeBatchTransferFrom(to, marketAddress, ids, amounts, data);
+    uint256[] memory totals = balanceOfBatch(owners, removals.values());
```

Comment:
This will grow linearly with the total supply in the market which is not going to work.  I think we're going to need to maintain running balances of these as we go.

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/structs/EnumerableSet.sol#L351-L354

Reply:
tracked